core/src/main/kotlin/com/exactpro/th2/read/db/impl/grpc/DataBaseReaderGrpcServer.kt

```diff
@@ -168,33 +170,59 @@ class DataBaseReaderGrpcServer(
     }
 
     private class GrpcExecuteListener(
-        private val observer: StreamObserver<QueryResponse>,
+        streamObserver: StreamObserver<QueryResponse>,
+        private val executionId: Long,
         private val event: Event,
         private val onEvent: (Event) -> Unit,
     ) : ResultListener {
+        private val skipped = AtomicLong()
+        private val observer = streamObserver as ServerCallStreamObserver<QueryResponse>
+
         override fun onRow(sourceId: DataSourceId, row: TableRow) {
-            requireNotNull(row.executionId) {
-                "'Execution id' is null for row: $row"
+            check(executionId == row.executionId) {
+                "'Execution id' isn't equal to '$executionId', row: $row"
+            }
+
+            while (!observer.isReady) {
+                if (observer.isCancelled) {
+                    skipped.incrementAndGet()
+                    return
+                }
+                Thread.yield()
+            }
+
+            try {
+                observer.onNext(
+                    QueryResponse.newBuilder()
+                        .putRows(row)
+                        .setExecutionId(row.executionId)
+                        .build()
+                )
+            } catch (e: StatusRuntimeException) {
+                LOGGER.error(e) { "Couldn't send next gRPC message for '$executionId' execution" }
+                skipped.incrementAndGet()
             }
-            observer.onNext(
-                QueryResponse.newBuilder()
-                    .putRows(row)
-                    .setExecutionId(row.executionId)
-                    .build()
-            )
         }
 
         override fun onError(error: Throwable) {
             observer.onError(error)
             event.endTimestamp()
                 .exception(error, true)
                 .also(onEvent)
+            logSkipped()
         }
 
         override fun onComplete() {
             observer.onCompleted()
             event.endTimestamp()
                 .also(onEvent)
+            logSkipped()
+        }
+
+        private fun logSkipped() {
+            if (skipped.get() > 0) {
+                LOGGER.warn { "gRPC consumer doesn't receives ${skipped.get()} rows for '$executionId' execution" }
+            }
```

Comment:
thought: I don't think we need this. If for some reason the connection was closed who cares how many rows it missed?